speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/FileDataStorage.cpp

```diff
@@ -98,64 +121,307 @@ void FileDataStorage::Shutdown()
+
+        if (partitionIndex == p_middle)
+        {
+            p_b = p_values[partitionIndex];
+            if (p_a != -1)
+            {
+                return;
+            }
+        }
+        else if (partitionIndex == p_middle - 1)
+        {
+            p_a = p_values[partitionIndex];
+            if (p_b != -1)
+            {
+                return;
+            }
+        }
+
+        if (partitionIndex >= p_middle)
+        {
+            GetMedianUtil(p_values, p_start, partitionIndex - 1, p_middle, p_a, p_b);
+        }
+        else
+        {
+            GetMedianUtil(p_values, partitionIndex + 1, p_end, p_middle, p_a, p_b);
+        }
     }
-    WRITE_STRING(m_outputStream, "NONE");
+}
+
+/// @brief partitions p_values around a randomly picked element
+/// @param p_values Array with values from the current compression step
+/// @param p_start start of the partition
+/// @param p_end end of the partition
+int FileDataStorage::RandomPartition(float* p_values, int p_start, int p_end)
+{
+    int pivot = m_random.NextInt(p_start, p_end + 1);
+    std::swap(p_values[pivot], p_values[p_end]);
+    return Partition(p_values, p_start, p_end);
+}
+
+/// @brief returns the correct position of the pivot element
+/// @param p_values Array with values from the current compression step
+/// @param p_start start of the partition
+/// @param p_end end of the partition
+int FileDataStorage::Partition(float* p_values, int p_start, int p_end)
+{
+    float lst = p_values[p_end];
+    int i = p_start, j = p_start;
+    while (j < p_end)
+    {
+        if (p_values[j] < lst)
+        {
+            std::swap(p_values[i], p_values[j]);
+            i++;
+        }
+        j++;
+    }
+    std::swap(p_values[i], p_values[j]);
+    return i;
+}
+
+/// @brief Get the least common value of the current compression step
+/// @param p_values Array with values from the current compression step
+/// @return The least common in an array for a variable
+int FileDataStorage::GetLeastCommon(int* p_values) const
+{
+    std::map<int, int> p_frequencies;
+
+    // list of all values in p_values
+    int p_valuesList[COMPRESSION_LIMIT];
+    int p_valueCount = 0;
+
+    // create a map from values in p_values to the corresponding frequencies
+    for (int i = 0; i < m_compressionRate; i++)
+    {
+        if (p_frequencies[p_values[i]])
+        {
+            p_frequencies[p_values[i]]++;
+        }
+        else
+        {
+            p_frequencies[p_values[i]] = 1;
+            p_valuesList[p_valueCount] = p_values[i];
+            p_valueCount++;
+        }
+    }
+
+    int p_leastCommon;
+    int p_minCount = m_compressionRate + 1;
+
+    // find the value with the lowest frequency
+    for (int i = 0; i < p_valueCount; i++)
+    {
+        if (p_frequencies[p_valuesList[i]] < p_minCount)
+        {
+            p_leastCommon = p_valuesList[i];
+            p_minCount = p_frequencies[p_valuesList[i]];
+        }
+    }
+
+    return p_leastCommon;
```

Comment:
Not paramaters, shouldn't have p_. Also aren't you returning a stack allocated value? 
```suggestion
int FileDataStorage::GetLeastCommon(int* p_values) const
{
    std::map<int, int> frequencies;

    // list of all values in p_values
    int valuesList[COMPRESSION_LIMIT];
    int valueCount = 0;

    // create a map from values in p_values to the corresponding frequencies
    for (int i = 0; i < m_compressionRate; i++)
    {
        if (frequencies[p_values[i]])
        {
            frequencies[p_values[i]]++;
        }
        else
        {
            frequencies[p_values[i]] = 1;
            valuesList[valueCount] = p_values[i];
            valueCount++;
        }
    }

    int leastCommon;
    int minCount = m_compressionRate + 1;

    // find the value with the lowest frequency
    for (int i = 0; i < valueCount; i++)
    {
        if (frequencies[valuesList[i]] < minCount)
        {
            leastCommon = valuesList[i];
            minCount = frequencies[valuesList[i]];
        }
    }

    return leastCommon;
```